### DISCUSSION TITLE
sub-dependencies passing parameters and returning results

### DISCUSSION THREAD

**Discussion by pincoin** [fastapi/fastapi#5053] [Questions] [answered]
Hello,

I have a small project with FastAPI and SQLAlchemy Core 1.4.

I've got my API function like this:

```py
async def do_something(
    ...
    current_user: dict = Depends(get_superuser),
    conn: AsyncConnection = Depends(engine_begin),
):
   ...
```

`do_something` API function depends on `get_superuser` and SQLAlchemy connection by `engine_begin`.

I am planning on sub-dependencies like this:

`do_something` -> `get_superuser` -> `get_current_user`

`get_current_user` dependency is not required to access database during JWT validation.

However, `get_superuser` is required to access database to check permissions or something else.

```py
async def get_superuser(
    current_user: dict = Depends(get_current_user),
    conn: AsyncConnection = Depends(engine_connect),
) -> dict:
    ...

    return {
        "superuser": row._mapping,
        "conn": conn,
    }
```

At first, I was able to make `get_superuser` depends on `get_current_user` and `engine_begin`, and returned dict as shown above.

Personally it does not make sense if `get_superuser` returns both superuser and database connection.

I tried to put the path decorator `dependencies=[Depends(engine_begin)]` Then, I could not get the connection handle.

Another solution might be double dependencies like `do_something` depends on `engine_begin` and `get_superuser` depends on `engine_begin`. I think it will cause another problem due to nested transaction begin.

I would like to know if there is a clear way to **check authorization/permission after authentication**.

I want to be able to pass `conn` from my `do_something` api to `get_superuser` dependency.

Thank you.

**Comment by zoliknemet** [comment, marked as answer]
@pincoin 
If I understood you correctly, you are missing the point of Depends.
They are cached by default. That means that you can call same dependency over and over and you will get same object.
Example:
```python
import uvicorn as uvicorn
from fastapi import FastAPI, Depends

app = FastAPI()


async def second_level_dep():
    print("second level")
    yield "second level string"
    print("second level over")


async def first_level_dep(second_level_dep=Depends(second_level_dep)):
    print("first level")
    yield f"first level stirng - {second_level_dep}"
    print("first level over")


@app.get("/do-somthing")
async def do_something(
    first_level: str = Depends(first_level_dep),
    first_level_again: str = Depends(first_level_dep),
    second_level: str = Depends(second_level_dep),
    second_level_again: str = Depends(second_level_dep),
    second_level_again_two: str = Depends(second_level_dep),
):
    print(f"result of first level is: {first_level}")


if __name__ == "__main__":
    uvicorn.run(app)
```

```text
second level
first level
result of first level is: first level stirng - second level string
first level over
second level over
```

second_level_dep is called only once per request so you are ok with calling this `conn: AsyncConnection = Depends(engine_connect)` when ever you need it, it sill make 1 connection/transaction and it will be fine

**Comment by pincoin** [comment]
Thanks to @zoliknemet 

I wrote my code like this:

SQLAlchemy engine connection dependency

```py
async def engine_connect() -> typing.Generator:
    async with engine.connect() as conn:
        yield conn
    await engine.dispose()
```

* `get_superuser` depends on `get_current_user` and `engine_connect`

```py
async def get_current_user(
    token: str = Depends(oauth2_scheme),
) -> dict:
    ...


async def get_superuser(
    current_user: dict = Depends(get_current_user),
    conn: AsyncConnection = Depends(engine_connect),
) -> dict:
    ...
```

* API function depends on `get_superuser` and `engine_connect`

```py
@router.put(
    "/groups/{group_id}",
)
async def update_group(
    group: GroupUpdate,
    group_id: int = Query(default=0, gt=0),
    superuser: dict = Depends(get_superuser),
    conn: AsyncConnection = Depends(engine_connect),
):
    ...
```

Now I've got only **ONE** engine connection because FastAPI guarantees calling dependency once, but I needed to build another context manager in order to prevent **"nested"** transactions.

```py
import contextlib

@contextlib.contextmanager
def transaction(connection):
    if not connection.in_transaction():
        with connection.begin():
            yield connection
    else:
        yield connection



# API function uses CRUD model class which has some codes like that:
    with transaction(connection):
        connection.execute(statement)
        connection.commit() # autocommit = False
```

Results
```
2022-06-19 15:37:38,738 INFO sqlalchemy.engine.Engine BEGIN (implicit)
2022-06-19 15:37:38,738 INFO sqlalchemy.engine.Engine SELECT auth_user.id, auth_user.password, auth_user.last_login, auth_user.is_superuser, auth_user.username, auth_user.first_name, auth_user.last_name, auth_user.email, auth_user.is_staff, auth_user.is_active, auth_user.date_joined 
FROM auth_user 
WHERE auth_user.id = %s AND auth_user.is_active = true AND auth_user.is_superuser = true
2022-06-19 15:37:38,738 INFO sqlalchemy.engine.Engine [cached since 14.44s ago] (1,)
2022-06-19 15:37:38,741 INFO sqlalchemy.engine.Engine SELECT auth_group.id, auth_group.name 
FROM auth_group 
WHERE auth_group.id = %s
2022-06-19 15:37:38,741 INFO sqlalchemy.engine.Engine [generated in 0.00021s] (17,)
2022-06-19 15:37:38,743 INFO sqlalchemy.engine.Engine UPDATE auth_group SET id=%s, name=%s WHERE auth_group.id = %s
2022-06-19 15:37:38,744 INFO sqlalchemy.engine.Engine [generated in 0.00021s] (17, 'string', 17)
2022-06-19 15:37:38,745 INFO sqlalchemy.engine.Engine COMMIT
```

I believe now I follow the coding rules and conventions by both FastAPI and SQLAlchemy.

Thank you.
